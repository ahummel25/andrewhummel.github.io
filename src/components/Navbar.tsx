import React, { FC, useEffect, useState } from 'react';
import styled from '@emotion/styled';
import Drawer from '@material-ui/core/Drawer';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import Divider from '@material-ui/core/Divider';

import { useWindowDimensions } from '../hooks';
import { NavbarProps, NavListProps } from '../interfaces';
import { colors, fonts } from '../styles/variables';

import WeatherCard from './WeatherCard';

type StyledLiProps = {
  color: string;
};

const DRAWER_WIDTH = 190;
const MAX_WINDOW_WIDTH = 992;

const DrawerContainer = styled(({ ...rest }) => <Drawer {...rest} />)`
  flex-shrink: 0;

  ul {
    margin: 0;
  }

  .name-title {
    h1 {
      color: ${colors.lightGreen};
      font-family: ${fonts.robotoSansSerif};
      font-weight: 100;
      font-size: 30.45px;
      margin: 0 0 5px 0;
    }

    span {
      font-size: 14px;
    }

    background: ${colors.brandBackground};
    display: inline-block;
    min-height: 100px;
    text-align: left;
    padding: 30px 0 30px 30px;
    width: ${DRAWER_WIDTH}px;
  }
`;

const AppBarCustom = styled(({ ...rest }) => <AppBar {...rest} />)`
  display: block;
  text-align: center;
  font-size: 22px;

  .name-title > a {
    color: rgb(0, 128, 115);
    display: block;
    margin-bottom: -7px;
    padding-top: 10px;
  }

  span {
    font-size: 14px;
    display: block;
    padding-bottom: 3px;
  }

  a > i {
    font-size: 35px;
    display: block;
    float: left;
    padding-top: 15px;
    margin-right: -30px;
  }

  .icon-green {
    color: rgb(0, 128, 115);
  }

  .tool-bar {
    display: block;
    text-align: center;
  }
`;

const DrawerLi = styled.li<StyledLiProps>`
  cursor: pointer;
  padding: 0;
  font-weight: 300;
  display: flex;
  align-items: center;

  @keyframes hover-on {
    from {
      background-color: rgb(239, 340, 241);
    }
    to {
      background-color: rgb(250, 250, 250);
    }
  }

  &:hover {
    border-left: 3px solid ${(props): string => props.color};
    animation: hover-on 0.75s;
    animation-fill-mode: forwards;
    will-change: opacity, transform;
  }

  a {
    color: rgb(51, 51, 51);
    display: flex;
    align-items: center;
    padding: 10px 0 8px 10px;
    width: 100%;
  }

  i {
    color: rgb(51, 51, 51);
    font-weight: 500;
    display: flex;
    align-items: center;
    padding: 0 10px 0 5px;
    &.icon-color {
      color: ${(props): string => props.color};
    }
  }

  .active {
    background-color: rgb(250, 250, 250);
    border-left: 3px solid ${(props): string => props.color};
    font-weight: 500;
  }
`;

const navItems = [
  {
    text: 'About',
    icon: {
      type: 'person',
      color: 'rgb(0, 128, 115)'
    },
    ref: 'about'
  },
  {
    text: 'Experience',
    icon: {
      type: 'trending_up',
      color: 'rgb(0, 188, 212)'
    },
    ref: 'experience'
  },
  {
    text: 'Skills',
    icon: {
      type: 'equalizer',
      color: 'rgb(156,38,176)'
    },
    ref: 'skills'
  },
  {
    text: 'Certifications',
    icon: {
      type: 'star',
      color: 'rgb(244,67,53)'
    },
    ref: 'certifications'
  },
  {
    text: 'Contact',
    icon: {
      type: 'mail',
      color: 'rgb(121, 85, 72)'
    },
    ref: 'contact'
  }
];

const RenderFullDrawer: FC<Record<string, unknown>> = (): JSX.Element => (
  <DrawerContainer variant="permanent" anchor="left">
    <div className="name-title">
      <a href="/">
        <h1>
          Andrew <br /> Hummel
        </h1>
        <span className="brown-text light">Full Stack Developer</span>
      </a>
    </div>
    <Divider />
    <NavBarLists />
    <WeatherCard />
  </DrawerContainer>
);

const RenderMobileDrawer: FC<Record<string, unknown>> = (): JSX.Element => {
  const [open, setOpen] = useState<boolean>(false);

  const toggleDrawer = (
    event: React.KeyboardEvent | React.MouseEvent
  ): void => {
    event.preventDefault();

    if (
      event.type === 'keydown' &&
      ((event as React.KeyboardEvent).key === 'Tab' ||
        (event as React.KeyboardEvent).key === 'Shift')
    ) {
      return;
    }

    setOpen(!open);
  };

  return (
    <>
      <AppBarCustom color="inherit" position="fixed">
        <Toolbar className="tool-bar">
          <a href="/" onClick={toggleDrawer}>
            <i className="material-icons icon-green">menu</i>
          </a>
          <div className="name-title">
            <a href="/">Andrew Hummel</a>
            <span className="brown-text light">Full Stack Developer</span>
          </div>
        </Toolbar>
      </AppBarCustom>
      <DrawerContainer anchor="left" open={open} onClose={toggleDrawer}>
        <div className="name-title">
          <a href="/">
            <h1>Andrew Hummel</h1>
            <span className="brown-text light">Full Stack Developer</span>
          </a>
        </div>
        <Divider />
        <NavBarLists setOpen={setOpen} />
        <WeatherCard />
      </DrawerContainer>
    </>
  );
};

const NavBarLists: FC<NavListProps> = ({ setOpen }): JSX.Element => (
  <ul>
    {navItems.map(
      ({ text, icon: { type, color }, ref }, index): JSX.Element => (
        <DrawerLi key={index} color={color}>
          <a
            href={`#${ref}`}
            className="waves-effect waves-dark"
            onClick={(): void => {
              if (setOpen) setOpen(false);
            }}
          >
            <i className="small material-icons icon-color">{type}</i>
            {text}
          </a>
        </DrawerLi>
      )
    )}
  </ul>
);

const Navbar: FC<NavbarProps> = ({ setDrawerWidth }): JSX.Element => {
  const { width } = useWindowDimensions();
  const [showMobileNav, setShowMobileNav] = useState<boolean>(false);

  useEffect((): void => {
    setDrawerWidth(DRAWER_WIDTH);
  }, [setDrawerWidth]);

  useEffect((): void => {
    setShowMobileNav(width < MAX_WINDOW_WIDTH);
  }, [width]);

  return <>{showMobileNav ? <RenderMobileDrawer /> : <RenderFullDrawer />}</>;
};

export default Navbar;
