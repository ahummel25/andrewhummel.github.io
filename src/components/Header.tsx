import React, { FC } from "react";
import styled from "@emotion/styled";
import Img from "gatsby-image";

import Navbar from "../components/Navbar";
import { useGetProfileImage } from "../hooks";
import { HeaderProps } from "../interfaces";

const ImgStyled = styled(({ ...rest }) => <Img {...rest} />)`
  @media (min-width: 992px) {
    margin-top: 0;
  }
  margin-top: 64px;
  display: inline-block;
`;

const Header: FC<HeaderProps> = ({
  refsToForward,
  setDrawerWidth
}): JSX.Element => {
  const imgData = useGetProfileImage();

  return (
    <>
      <Navbar refsToForward={refsToForward} setDrawerWidth={setDrawerWidth} />
      <ImgStyled
        fluid={imgData.file.childImageSharp.fluid}
        alt="Profile Shot"
      />
    </>
  );
};

export default Header;
