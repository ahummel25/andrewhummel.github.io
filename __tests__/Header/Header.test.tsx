import React from 'react';
import renderer from 'react-test-renderer';

import Header from '../../src/components/Header';
import Navbar from '../../src/components/Navbar';

describe('Header', () => {
  const setDrawerWidth = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    const tree = renderer.create(<Header setDrawerWidth={setDrawerWidth} />);

    expect(tree).toBeDefined();
  });

  it('has correct props for subcomponent and HTML element styles', () => {
    const tree = renderer.create(<Header setDrawerWidth={setDrawerWidth} />);
    const root = tree.root;

    expect(root.findByType(Navbar).props.setDrawerWidth).toBe(setDrawerWidth);

    const section = root.findAll(el => el.type === 'section');
    expect(section).toHaveLength(1);
    expect(section[0].props.style).toStrictEqual({ display: 'contents' });
  });
});
