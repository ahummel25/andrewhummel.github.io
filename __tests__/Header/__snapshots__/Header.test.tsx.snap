// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders correctly 1`] = `
.emotion-0 {
  display: block;
  text-align: center;
  font-size: 22px;
}

.emotion-0 .name-title>a {
  color: rgb(0, 128, 115);
  display: block;
  margin-bottom: -7px;
  padding-top: 10px;
}

.emotion-0 span {
  font-size: 14px;
  display: block;
  padding-bottom: 3px;
}

.emotion-0 a>i {
  font-size: 35px;
  display: block;
  float: left;
  padding-top: 15px;
  margin-right: -30px;
}

.emotion-0 .icon-green {
  color: rgb(0, 128, 115);
}

.emotion-0 .tool-bar {
  display: block;
  text-align: center;
}

.emotion-1 {
  margin-top: 64px;
  display: inline-block;
}

@media (max-width: 576px) {
  .emotion-1 {
    margin-top: 60px;
  }
}

@media (min-width: 992px) {
  .emotion-1 {
    margin-top: 0;
  }
}

<section
  className="section scrollspy"
  id="header"
  style={
    Object {
      "display": "contents",
    }
  }
>
  <header
    className="MuiPaper-root MuiAppBar-root MuiAppBar-positionFixed MuiAppBar-colorInherit emotion-0 mui-fixed MuiPaper-elevation4"
  >
    <div
      className="MuiToolbar-root MuiToolbar-regular tool-bar MuiToolbar-gutters"
    >
      <a
        href="/"
        onClick={[Function]}
      >
        <i
          className="material-icons icon-green"
        >
          menu
        </i>
      </a>
      <div
        className="name-title"
      >
        <a
          href="/"
        >
          Andrew Hummel
        </a>
        <span
          className="brown-text light"
        >
          Full Stack Developer
        </span>
      </div>
    </div>
  </header>
  <div
    className="emotion-1 gatsby-image-wrapper"
    style={
      Object {
        "maxHeight": null,
        "maxWidth": null,
        "overflow": "hidden",
        "position": "relative",
      }
    }
  >
    <div
      aria-hidden={true}
      style={
        Object {
          "paddingBottom": "63.83896477354421%",
          "width": "100%",
        }
      }
    />
    <img
      alt=""
      aria-hidden={true}
      src="data:image/jpeg;base64,/9j/2wBDABALDA4MChAODQ4SERATGCgaGBYWGDEj"
      style={
        Object {
          "height": "100%",
          "left": 0,
          "objectFit": "cover",
          "objectPosition": "center",
          "opacity": 1,
          "position": "absolute",
          "top": 0,
          "transitionDelay": "500ms",
          "width": "100%",
        }
      }
    />
    <picture>
      <source
        sizes="(max-width: 800px) 100vw, 800px"
        srcSet="/static/a68ec0a3134b8610c476cd0fa3c0cf7e/fd013/profile.jpg 200w,
						  /static/a68ec0a3134b8610c476cd0fa3c0cf7e/25252/profile.jpg 400w,
						  /static/a68ec0a3134b8610c476cd0fa3c0cf7e/2f1b1/profile.jpg 800w,
						  /static/a68ec0a3134b8610c476cd0fa3c0cf7e/0ff54/profile.jpg 1200w,
						  /static/a68ec0a3134b8610c476cd0fa3c0cf7e/6d185/profile.jpg 1391w"
      />
      <img
        alt="Profile Shot"
        loading="lazy"
        onLoad={[Function]}
        sizes="(max-width: 800px) 100vw, 800px"
        src="/static/a68ec0a3134b8610c476cd0fa3c0cf7e/2f1b1/profile.jpg"
        srcSet="/static/a68ec0a3134b8610c476cd0fa3c0cf7e/fd013/profile.jpg 200w,
						  /static/a68ec0a3134b8610c476cd0fa3c0cf7e/25252/profile.jpg 400w,
						  /static/a68ec0a3134b8610c476cd0fa3c0cf7e/2f1b1/profile.jpg 800w,
						  /static/a68ec0a3134b8610c476cd0fa3c0cf7e/0ff54/profile.jpg 1200w,
						  /static/a68ec0a3134b8610c476cd0fa3c0cf7e/6d185/profile.jpg 1391w"
        style={
          Object {
            "height": "100%",
            "left": 0,
            "objectFit": "cover",
            "objectPosition": "center",
            "opacity": 0,
            "position": "absolute",
            "top": 0,
            "transition": "opacity 500ms",
            "width": "100%",
          }
        }
      />
    </picture>
    <noscript
      dangerouslySetInnerHTML={
        Object {
          "__html": "<picture><source srcset=\\"/static/a68ec0a3134b8610c476cd0fa3c0cf7e/fd013/profile.jpg 200w,
						  /static/a68ec0a3134b8610c476cd0fa3c0cf7e/25252/profile.jpg 400w,
						  /static/a68ec0a3134b8610c476cd0fa3c0cf7e/2f1b1/profile.jpg 800w,
						  /static/a68ec0a3134b8610c476cd0fa3c0cf7e/0ff54/profile.jpg 1200w,
						  /static/a68ec0a3134b8610c476cd0fa3c0cf7e/6d185/profile.jpg 1391w\\" sizes=\\"(max-width: 800px) 100vw, 800px\\" /><img loading=\\"lazy\\" sizes=\\"(max-width: 800px) 100vw, 800px\\" srcset=\\"/static/a68ec0a3134b8610c476cd0fa3c0cf7e/fd013/profile.jpg 200w,
						  /static/a68ec0a3134b8610c476cd0fa3c0cf7e/25252/profile.jpg 400w,
						  /static/a68ec0a3134b8610c476cd0fa3c0cf7e/2f1b1/profile.jpg 800w,
						  /static/a68ec0a3134b8610c476cd0fa3c0cf7e/0ff54/profile.jpg 1200w,
						  /static/a68ec0a3134b8610c476cd0fa3c0cf7e/6d185/profile.jpg 1391w\\" src=\\"/static/a68ec0a3134b8610c476cd0fa3c0cf7e/2f1b1/profile.jpg\\" alt=\\"Profile Shot\\" style=\\"position:absolute;top:0;left:0;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center\\"/></picture>",
        }
      }
    />
  </div>
</section>
`;
